### ISSUE TITLE
Remove `@spell` for strings ?

### ISSUE THREAD

**Issue by ShellCode33** [nvim-treesitter/nvim-treesitter#5202]
Hey ! This will basically be a copy-pasta from an issue I filed upstream (See https://github.com/neovim/neovim/issues/24585).

### Problem

In my opinion the spell checker reports too many spelling issues for bash and lua (especially bash where you have to quote everything to prevent globbing). May I suggest removing `@spell` for strings in [runtime/queries/bash/highlights.scm](https://github.com/nvim-treesitter/nvim-treesitter/blob/2051c8603d572c5a0b23225549fd7d735adf115f/queries/bash/highlights.scm#L48) and [runtime/queries/lua/highlights.scm](https://github.com/nvim-treesitter/nvim-treesitter/blob/2051c8603d572c5a0b23225549fd7d735adf115f/queries/lua/highlights.scm#L244) ? It's not set for C and Python and my experience with the spell checker was way more pleasing !

I'd be glad to provide a PR for it but I have no understanding of Treesitter whatsoever... I'd rather let someone who knows what he's doing do it

### Expected behavior

No undercurls in the code :)

![image](https://github.com/neovim/neovim/assets/8455652/e79e3336-fb17-4673-a3d5-d579e6a942cb)

**Comment by clason** [comment]
Would be good to have a general discussion on which captures should by default have spell set (which should be a minimal set; i.e., if in doubt, don't).

**Comment by ShellCode33** [comment]
My opinion on this is that it should only be enabled for comments (exception being made for markdown and other note-related formats).

I do realise that sometimes you write proper sentences in strings that are worth spellchecking, but I feel like most of the time this is not what you want and should not be the default behavior.

Here's another example where spell checking is irrelevantly enabled for string literals:

![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/8455652/25181c6a-d47e-4725-b667-e24fb39253b4)

Might be worth highlighting that the docstring keywords are being processed by the spell checker as well. But I guess there's not much that can be done about it, and I think that's fine.

**Comment by clason** [comment]
Yes, that's the problem. There's not enough semantic information to enable this (efficiently) in a more targeted fashion. Which means, no matter what you do, somebody complains.

**Comment by ShellCode33** [comment]
> There's not enough semantic information to enable this (efficiently) in a more targeted fashion

That was indeed my understanding, this is why I totally agree with your previous statement :

> which should be a minimal set; i.e., if in doubt, don't

**Comment by lewis6991** [comment]
For reference here's a list of all the languages that currently spell check strings (it isn't many):

- `ada`
- `bash`
- `beancount`
- `chatito (text)`
- `clojure`
- `commonlisp`
- `ecma`
- `elixir`
- `erlang`
- `fennel`
- `fsh`
- `go`
- `haskell`
- `html_tags`
- `janet_simple`
- `json`
- `lua(u)`
- `php`
- `poe_filter`
- `racket`
- `ruby`
- `scheme`
- `verilog`
- `zig`
- `python` (only docstrings)

It's worth noting some query files have *smart* queries for spell checking strings. e.g:
- only spell check raw/literal strings that have no interpolation
- only spell check strings used for documentation (docstrings).

Longer term I think the solution is to add `@spell.string` and provide control to enable/disable spellchecking on these captures. A `vim.g/b.` variable should suffice for this.

Short term, we need to admit that not all strings should be treated equally, so we need to provide some guidelines on what kinds of strings should be spell checked #5203. We should also err on the side of removing spell checking for strings for a language if it can cause too many false-negative spelling errors (.e.g bash).

**Comment by lewis6991** [comment]
> Might be worth highlighting that the docstring keywords are being processed by the spell checker as well. But I guess there's not much that can be done about it, and I think that's fine.

We can probably fix it so certain semantic highlights also apply `@nospell`.

**Comment by ShellCode33** [comment]
A few comments on the ones I know:

- `bash` : as stated by Lewis, too many false positives due to the fact that quotes must be used everywhere for everything in bash
- `go` : debatable, kind of the same as `c` where `@spell` is not set (inconsistency), IMO we should remove it for string literals
- `html_tags` doesn't seem to apply to all strings, only to `text` sections, if I'm not mistaken this only applies to the inner portion of tags such as `<h1>Text</h1>`, so that should be fine
- `json` : mostly used for data storage, I don't think enabling the spell checker for it is relevant
- `python` : I think it's fine the way it is, it's a good balance between usefulness and usability

------------

> We can probably fix it so certain semantic highlights also apply @nospell.

Usually there's not a standard way of writing docstrings, let's take Python for example, you can write docstrings using `Epytext`, `reST`, `Google`, `Numpydoc`, etc. It would mean adding semantic support for all those variants if we want to be be able to spell check properly (and do the same for all the languages!). Which is not feasible in a reasonable time frame in my opinion. So the focus should not be on what's *inside* the strings/comments but rather if we should enable spell checking for them as a whole or not.

As a general rule of thumb, I would say the spell checker should be enabled by default for:

- Notes-specific file formats (e.g. Markdown)
- Comments
- Strings that are explicitly marked as documentation by some sort of "smart" queries

Everything else should not be spell checked.

**Comment by ShellCode33** [comment]
Another idea:

We could explicitly mark **everything** as `@spell` or `@nospell` when there's no doubt about it, and don't specify anything when it's ambiguous (e.g. string literals). That way it would allow us to have an additional opt-in parameter for ambiguous situations, for example:

```
vim.opt.spelloptions:append({ "ambiguous" })
```

**Comment by lucario387** [comment]
Should I close this, or move this to discussion?

**Comment by lewis6991** [comment]
I think we can close for now.

**Comment by clason** [comment]
Need to document somewhere (contributing.md) still.
